msal_extensions/persistence.py

```diff
@@ -0,0 +1,225 @@
+
+    @abc.abstractmethod
+    def save(self, content):
+        # type: (str) -> None
+        """Save the content into this persistence"""
+        raise NotImplementedError
+
+    @abc.abstractmethod
+    def load(self):
+        # type: () -> str
+        """Load content from this persistence"""
+        raise NotImplementedError
+
+    @abc.abstractmethod
+    def getmtime(self):
+        """Get the latest mtime of this persistence"""
+        raise NotImplementedError
+
+    @abc.abstractmethod
+    def get_location(self):
+        """Return the file path which this persistence stores (meta)data into"""
+        raise NotImplementedError
+
+
+class FilePersistence(BasePersistence):
+    """A generic persistence, storing data in a plain-text file"""
+
+    def __init__(self, location):
+        self._location = os.path.expanduser(location)
+        _mkdir_p(os.path.dirname(self._location))
+
+    def save(self, content):
+        # type: (str) -> None
+        """Save the content into this persistence"""
+        with open(self._location, 'w+') as handle:
+            handle.write(content)
+
+    def load(self):
+        # type: () -> str
+        """Load content from this persistence"""
+        with open(self._location, 'r') as handle:
+            return handle.read()
+
+    def getmtime(self):
+        """Get the latest mtime of this persistence"""
+        return os.path.getmtime(self._location)
+
+    def get_location(self):
+        return self._location
+
+
+class FilePersistenceWithDataProtection(FilePersistence):
+    """A generic persistence with data stored in a file,
+    protected by Win32 encryption APIs on Windows"""
+    is_encrypted = True
+
+    def __init__(self, location, entropy=''):
+        """Initialization could fail due to unsatisfied dependency"""
+        # pylint: disable=import-outside-toplevel
+        from .windows import WindowsDataProtectionAgent
+        self._dp_agent = WindowsDataProtectionAgent(entropy=entropy)
+        super(FilePersistenceWithDataProtection, self).__init__(location)
+
+    def save(self, content):
+        super(FilePersistenceWithDataProtection, self).save(
+            self._dp_agent.protect(content))
+
+    def load(self):
+        return self._dp_agent.unprotect(
+            super(FilePersistenceWithDataProtection, self).load())
+
+
+class KeychainPersistence(BasePersistence):
+    """A generic persistence with data stored in,
+    and protected by native Keychain libraries on OSX"""
+    is_encrypted = True
+
+    def __init__(self, location, service_name, account_name, dummy=" "):
+        """Initialization could fail due to unsatisfied dependency"""
+        if not (service_name and account_name):  # It would hang on OSX
+            raise ValueError("service_name and account_name are required")
+        from .osx import Keychain  # pylint: disable=import-outside-toplevel
+        self._file_persistence = FilePersistence(location)  # Favor composition
```

Comment:
I can understand your point. The `location` was named this way, and even at the first position, only because historically the entire persistence family were subclasses of `FilePersistence`. Given that it is no longer the case, sure it is good idea to at least rename `location` to something like `flag_path`? Naming is hard.

But then your other message saying your attempting `location=None` usage is an even bigger topic. Normally a required parameter is not expected to accept `None` unless it was documented otherwise. But did you hint that you foresee a usage to skip that flag file all together? I'm fine to choose to allow `flag_path=None` in this API surface (or to allow it implicitly inside implementation), but it would also mean the burden is on your side (i.e. Azure SDK) to remember to provide a flag file in order for cross-process timestamp to work. Is that what you want?

Reply:
What I meant regarding `location=None` (really, `not location`) is something simpler: you raise `ValueError` when `service_name` or `account_name` is falsey, to prevent a hang. Maybe it's worth doing the same for `location`, to get a better error message?

Reply:
Renamed it to `signal_location`. Also, a `signal_location = None` input would trigger a ValueError exception this time.

PS: The term "signal" did not come up from nowhere. During some of early days brainstorming, yet another file named "your_token_cache.bin.signal" was briefly considered, alongside with the developer-provided "your_token_cache.bin" and the auto-generated "your_token_cache.bin.lock". Although we ended up not explicitly maintaining such 3-files setup, the "your_token_cache.bin" provided by app developer will indeed be used as the signal file, when running on OSX and Linux Desktop.